docs/github-actions-workflow/v0.1/example.json

```
@@ -5,36 +5,27 @@
             "buildType": "https://slsa.dev/github-actions-workflow/v0.1?draft",
             "externalParameters": {
                 "inputs": {
-                    "mapValue": {
-                        "build_id": "123456768",
-                        "deploy_target": "deployment_sys_1a",
-                        "perform_deploy": "true"
-                    }
+                    "build_id": 123456768,
+                    "deploy_target": "deployment_sys_1a",
+                    "perform_deploy": "true"
                 },
-                "source": {
-                    "artifactRef": {
-                        "uri": "git+https://github.com/octocat/hello-world@refs/heads/main",
-                        "digest": { "sha1": "c27d339ee6075c1f744c5d4b200f7901aad2c369" }
-                    }
+                "source": "git+https://github.com/octocat/hello-world@refs/heads/main",
+                "vars": {
+                    "MASCOT": "Mona"
                 },
-                "inputs": {
-                    "mapValue": {
-                        "mascot": "Mona"
-                    }
-                },
-                "workflowPath": {
-                    "scalarValue": ".github/workflow/release.yml"
-                }
+                "workflow": ".github/workflow/release.yml"
```

There's a new claim in the GitHub Actions OIDC token: `workflow_ref` which has the following format (and always points to the triggering/initiating workflow):

```suggestion
                "workflow": "octocat/hello-world/.github/workflow/release.yml@refs/heads/main"
```

---

Another thought, we also expose `workflow_sha`, which will be different to the materials digest sha when using `pull_request_target` in github actions. Using this event to release/publish is an anti-pattern but seems good to capture nonetheless so we can spot this pattern.

What if we changed this to an object?
```suggestion
                "workflow": {
                  "uri": "git+https://github.com/octocat/hello-world/.github/workflow/release.yml",
                  "digest": { "sha1": "c27d339ee6075c1f744c5d4b200f7901aad2c369" }
                }
```

Looking ahead, I could see a world where it will be possible to reference initiating workflows from a different repo to the source with some kind of org policy, at which point we will need to track the uri and digest (btw, this isn't in the works as far as I know).

---

+1000! I wonder if this then falls into the pattern of referencing a complete `ArtifactReference` like in `resolvedDependencies` (I'm not sure this is a dependency though, but I like the idea of resolving the ref int oa digest)

---

What is the actual external interface to GitHub Actions? That is what I'm trying to capture here. As I mentioned in the other thread, the sha is not part of that interface AFAICT so it would go in `resolvedDependencies` instead.

Can the `github.repo` (called `source` here) ever differ from the the top-level workflow (aside from reusable workflows)? Does this happen for `pull_request_target`, or are you just saying it might happen in the future?

We could split this in two, though it would add a bit of complexity:

```jsonc
"source": "git+https://github.com/repo/repo@refs/heads/main",  // maybe omit this if the same as `workflow.repo`?
"workflow": {
  "repo": "git+https://github.com/workflow/workflow@refs/heads/main",
  "path": ".github/workflow/release.yml"
}
```

Regarding the format of the URI and path, my thinking was this:

- The `uri` is the locator of the git commit. It contains both the repository location (https://github.com/octocat/hello-world) and the ref that was evaluated (refs/heads/main). We use a single URI here so that we can join `externalParameters` and `resolvedDependencies`.
- The `workflow` is the path to the workflow within that commit. It seems redundant and error-prone to repeat parts of the `uri` here if it always has to agree. Alternatively, if it can ever disagree, then the workflow actually came from a different commit, and we should have a separate URI for that commit (again, for joining purposes).

---

Note that for a reusable workflow, the `workflow` is actually not relevant and we could omit it entirely, ie only report it as part of `systemParameters`.

---

Even if `workflow` would be irrelevant when using a reusable workflow, I think it make sense to keep it in `externalParameters` for consistency? I would think that it's up the verifier to expect certain values. 
When using a reusable workflow, that would go into the `reseolvedDependencies` right? We can get the hash via the `job_workflow_sha` claim from the id token.

---

PTAL. I just updated the schema as follows and I hope it is more understandable now.

1. Only support the specific event types that have clear semantics that align with the model. There are so many different event types, and new ones will crop up, that I think it's safer to take an allowlist approach. That avoids issues like `pull_request` and `pull_request_trigger` that seem error prone.
2. Rename to `workflowPath`.
3. Clarify documentation.

I originally tried to expand based on the idea of having a separate URI for workflow, particularly to support reusable workflows, but it made everything much harder to understand.

---

Explicitly dropping support for `pull_request` and especially `pull_request_target` is IMHO a very good idea.

---

👍  on only supporting a subset of event types.

What are your thoughts on aligning the value with what you get from the current claim: `workflow_ref`/`GITHUB_WORKFLOW_REF` env var so we don't need to restructure it?

`workflow_ref` has the following format: `octocat/hello-world/.github/workflow/release.yml@refs/heads/main`. Should we align it to the `source` URI format?

```
"workflow": "git+https://github.com/octocat/hello-world/.github/workflow/release.yml@refs/heads/main"
or
"workflowRef": "octocat/hello-world/.github/workflow/release.yml@refs/heads/main"
```

Although looks like [spdx](https://spdx.github.io/spdx-spec/v2.3/package-information/#773-examples) has a different idea about how this should be formatted.. `git+https://github.com/octocat/hello-world@master#.github/workflow/release.yml`

---

Ah just spotted https://github.com/slsa-framework/slsa/pull/582/files#diff-bbdc17f357d443804cf506038eba199aed6be071c25af1b7307e699bf9ad69c1R104 - seems doable but fwiw I would definitely prefer just copying in the `workflow_ref` in here for ease of use 😄

---

In my opinion, copying `workflow_ref` makes it easier to *generate* the provenance but harder to *verify* it. We only need to implement generation once or twice, while we need to implement verification and write down expectations many times. Therefore I think we should be optimizing for verification, not generation.

In particular, I'm worried about having duplicate information in `workflow_ref` that is already present in `source` because either it makes expectations be more verbose than necessary or it bakes in an assumption that they're always the same, which could change in the future.

I imagine we have three basic options:

- Three fields: `repo`, `ref`, `workflowPath`
- Two fields: `source` URI (repo@ref) and `workflowPath`
- One field: `workflow` URI (repo@ref#path)

I originally went with the two-field option because that matched our original idea of always using URIs to identify commits, and because you can join with `resolvedDependencies` using the same URI. But maybe that's not the right thing to optimize for. It does require parsing a URI even for expectations, since people will probably say "allow this particular repo but at any tag".

The one-field option is kind of nice since it's just a URI, but probably harder to work with in practice. It doesn't map 1:1 with `resolvedDependencies` (need to strip off the path, since the dependency is the full commit, not just that one file).

So maybe the three-field option is the best to go with? What do you think of that?

---

After more thought, I split into three fields, grouped under a common `workflow` object. Please take a look.

---

Good call out on verification! I like the three field workflow object, I think it will be easier to work with and clear how they are pointing to the same source now 👍